### ISSUE TITLE
multipart_extract: Unable to find multipart download range

### ISSUE THREAD

**Issue by sbabic** [zchunk/zchunk#21]
Hi,

(tag :1.1.3)

I am getting error from zck_dl and it is unable to create the zck file. I have two files (they are two ext4 filesystems generated with OE), just a few files differ on the two filesystem. The size is ~640MB. I built the two zck files and the new one is put on a webserver.
Then I run:

 zckdl -s  old.rootfs.ext4.zck http://localhost/swtest/new.ext4.zck

and it stops with :

Missing chunks: 765
Tried downloading too many ranges, reducing to 127
multipart_extract: Unable to find multipart download range
Download failed: Failed writing received data to disk/application
Unable to find multipart download range

Running in verbose mode,  it does not report much more:
[snip]
validate_chunk: Expected chunk checksum:   c2eae2e5a9a754726b190aa270cb5a22
validate_chunk: Calculated chunk checksum: c2eae2e5a9a754726b190aa270cb5a22
validate_chunk: Chunk 12068's checksum: valid
hash_init: Initializing SHA-512 hash
validate_checksums: No more data
validate_chunk: Expected chunk checksum:   f3a4b9b170efffe21d0c45841b192164
validate_chunk: Calculated chunk checksum: c2eae2e5a9a754726b190aa270cb5a22
validate_chunk: Chunk 12069's checksum: FAILED
hash_init: Initializing SHA-256 hash
Missing chunks: 765
Tried downloading too many ranges, reducing to 127
multipart_get_boundary: Multipart boundary: 4385da70938e99a5
multipart_extract: Unable to find multipart download range
Download failed: Failed writing received data to disk/application
Unable to find multipart download range
comp_close: Closing compression
comp_close: Closing compression

The two zck look ok if I run zck_read_header:

zck_read_header new.ext4.zck

Overall checksum type: SHA-256
Header size: 254642
Header checksum: d8033e42f225ba0447256533a1d99aeda8ea7c88b4e91ecf435d3fca8c6c216b
Data size: 197436855
Data checksum: b0cf95d38bababfc5752742909a889705937b04cb0d8b2782bbedf6fb599a1ed
Chunk count: 12070
Chunk checksum type: SHA-512/128

and 

zck_read_header old.ext4.zck

Overall checksum type: SHA-256
Header size: 254780
Header checksum: bcc383938bf45404e82ccf05b745398616aeb24e870a766d4636b339f6163a07
Data size: 197343471
Data checksum: 975616effd3070692956926c2a1682bfb0a5a2c6a8c2cb6904d908d43737f360
Chunk count: 12077
Chunk checksum type: SHA-512/128

 zck_delta_size reports a very low delta, no idea if this is correct, but there is no error:

hash_setup: Setting up hash type SHA-256
read_lead: Setting header and full digest hash type to SHA-256
read_lead: Reading header digest
read_lead: Parsed lead: 41 bytes
read_header_from_file: Reading the rest of the header: 254601 bytes
validate_header: Checking header checksum
validate_header: Expected header checksum:   d8033e42f225ba0447256533a1d99aeda8ea7c88b4e91ecf435d3fca8c6c216b
validate_header: Calculated header checksum: d8033e42f225ba0447256533a1d99aeda8ea7c88b4e91ecf435d3fca8c6c216b
validate_header: Header checksum valid
read_preface: Reading data digest
read_preface: Reading compression type and index size
set_comp_type: Setting compression to zstd
comp_init: Initializing zstd compression
read_index: Reading index
set_chunk_hash_type: Setting chunk hash to SHA-512/128
hash_setup: Setting up hash type SHA-512/128
read_sig: Signature count: 0
hash_setup: Setting up hash type SHA-256
read_lead: Setting header and full digest hash type to SHA-256
read_lead: Reading header digest
read_lead: Parsed lead: 41 bytes
read_header_from_file: Reading the rest of the header: 254601 bytes
validate_header: Checking header checksum
validate_header: Expected header checksum:   d8033e42f225ba0447256533a1d99aeda8ea7c88b4e91ecf435d3fca8c6c216b
validate_header: Calculated header checksum: d8033e42f225ba0447256533a1d99aeda8ea7c88b4e91ecf435d3fca8c6c216b
validate_header: Header checksum valid
read_preface: Reading data digest
read_preface: Reading compression type and index size
set_comp_type: Setting compression to zstd
comp_init: Initializing zstd compression
read_index: Reading index
set_chunk_hash_type: Setting chunk hash to SHA-512/128
hash_setup: Setting up hash type SHA-512/128
read_sig: Signature count: 0
Would download 254642 of 197691497 bytes
Matched 12070 of 12070 chunks
comp_close: Closing compression
comp_close: Closing compression

Any idea what can I look or if I am doing something clearly wrong ?

Thanks,
Stefano

**Comment by jdieter** [comment]
The zchunk files you've created are fine.

It looks like zckdl is unable to find the boundary used by the web server to separate the chunks.  Which web server are you using?

**Comment by sbabic** [comment]
It is apache2, no special configuration, distro is Ubuntu 18.04 (apache 2.4.29).

**Comment by sbabic** [comment]
It seems I have found a reason. There is a very old post regarding DOS on Apache with multiple Range:
https://lwn.net/Articles/456723/

I checked with Wireshark and apache answers just with the first Range, and ignores the other ones. This can be easy tested outside zchunk with curl:
curl  http://localhost/new.zck -H "Range: bytes=254642-462122,468195-474664,1892209-1914258,2393709-2408176,2413019-2432525,2468697-2488068,2498892-2510633" --output mess
(The list is the same sent by zckdl). Just the first range is answered by the Webserver. In fact, if I force zckdl to use a single range with:
diff --git a/src/zck_dl.c b/src/zck_dl.c
index 045f08f..6b80c70 100644
--- a/src/zck_dl.c
+++ b/src/zck_dl.c
@@ -58,10 +58,6 @@ static struct argp_option options[] = {
 };
 
 static int range_attempt[] = {
-    255,
-    127,
-    7,
-    2,
     1
 };
 
Then it works as expected and the .zck is created correctly. So I do not know if Multiple Range is a feature that is dropped by many Webservers for security reason or can be enabled in some way (I have not found anything in apache2 configuration).

**Comment by jdieter** [comment]
Do you mind tarring up your apache2 configuration and sending it to me (obviously without any keys or certs)?

I am aware of the link you posted above, and most webservers have a default limit of 256 ranges in a single request, a value that can be configured in Apache.  What's unusual here is the way the webserver is responding when too many ranges are requested.  Normally zck_dl gets an error, which then causes it to drop to the next range_attempt and then the next (with a maximum of four failures getting you from 255 ranges down to one).

If this is how Ubuntu is now handling too many range requests, then I need to come up with a way of working around it.

**Comment by jdieter** [comment]
I've just run a test with a clean 18.04 install and everything works correctly, so it looks like this is a site-specific setting:
```
$ zckdl -s linux-4.19.80.tar.zck http://localhost:8080/linux-4.19.83.tar.zck
Missing chunks: 12399
Tried downloading too many ranges, reducing to 127
Downloaded 138087457 bytes
```

If you didn't mind sending me your apache2 configuration, I would like to see if I can make zckdl handle it properly.

**Comment by sbabic** [comment]
Interesting. I tested on two servers, both with 18.04, getting the same errors. This is my apache2.conf:

# This is the main Apache server configuration file.  It contains the
# configuration directives that give the server its instructions.
# See http://httpd.apache.org/docs/2.4/ for detailed information about
# the directives and /usr/share/doc/apache2/README.Debian about Debian specific
# hints.
#
#
# Summary of how the Apache 2 configuration works in Debian:
# The Apache 2 web server configuration in Debian is quite different to
# upstream's suggested way to configure the web server. This is because Debian's
# default Apache2 installation attempts to make adding and removing modules,
# virtual hosts, and extra configuration directives as flexible as possible, in
# order to make automating the changes and administering the server as easy as
# possible.

# It is split into several files forming the configuration hierarchy outlined
# below, all located in the /etc/apache2/ directory:
#
#	/etc/apache2/
#	|-- apache2.conf
#	|	`--  ports.conf
#	|-- mods-enabled
#	|	|-- *.load
#	|	`-- *.conf
#	|-- conf-enabled
#	|	`-- *.conf
# 	`-- sites-enabled
#	 	`-- *.conf
#
#
# * apache2.conf is the main configuration file (this file). It puts the pieces
#   together by including all remaining configuration files when starting up the
#   web server.
#
# * ports.conf is always included from the main configuration file. It is
#   supposed to determine listening ports for incoming connections which can be
#   customized anytime.
#
# * Configuration files in the mods-enabled/, conf-enabled/ and sites-enabled/
#   directories contain particular configuration snippets which manage modules,
#   global configuration fragments, or virtual host configurations,
#   respectively.
#
#   They are activated by symlinking available configuration files from their
#   respective *-available/ counterparts. These should be managed by using our
#   helpers a2enmod/a2dismod, a2ensite/a2dissite and a2enconf/a2disconf. See
#   their respective man pages for detailed information.
#
# * The binary is called apache2. Due to the use of environment variables, in
#   the default configuration, apache2 needs to be started/stopped with
#   /etc/init.d/apache2 or apache2ctl. Calling /usr/bin/apache2 directly will not
#   work with the default configuration.


# Global configuration
#

#
# ServerRoot: The top of the directory tree under which the server's
# configuration, error, and log files are kept.
#
# NOTE!  If you intend to place this on an NFS (or otherwise network)
# mounted filesystem then please read the Mutex documentation (available
# at <URL:http://httpd.apache.org/docs/2.4/mod/core.html#mutex>);
# you will save yourself a lot of trouble.
#
# Do NOT add a slash at the end of the directory path.
#
#ServerRoot "/etc/apache2"

#
# The accept serialization lock file MUST BE STORED ON A LOCAL DISK.
#
#Mutex file:${APACHE_LOCK_DIR} default

#
# The directory where shm and other runtime files will be stored.
#

DefaultRuntimeDir ${APACHE_RUN_DIR}

#
# PidFile: The file in which the server should record its process
# identification number when it starts.
# This needs to be set in /etc/apache2/envvars
#
PidFile ${APACHE_PID_FILE}

#
# Timeout: The number of seconds before receives and sends time out.
#
Timeout 300

#
# KeepAlive: Whether or not to allow persistent connections (more than
# one request per connection). Set to "Off" to deactivate.
#
KeepAlive On

#
# MaxKeepAliveRequests: The maximum number of requests to allow
# during a persistent connection. Set to 0 to allow an unlimited amount.
# We recommend you leave this number high, for maximum performance.
#
MaxKeepAliveRequests 100

#
# KeepAliveTimeout: Number of seconds to wait for the next request from the
# same client on the same connection.
#
KeepAliveTimeout 5


# These need to be set in /etc/apache2/envvars
User ${APACHE_RUN_USER}
Group ${APACHE_RUN_GROUP}

#
# HostnameLookups: Log the names of clients or just their IP addresses
# e.g., www.apache.org (on) or 204.62.129.132 (off).
# The default is off because it'd be overall better for the net if people
# had to knowingly turn this feature on, since enabling it means that
# each client request will result in AT LEAST one lookup request to the
# nameserver.
#
HostnameLookups Off

# ErrorLog: The location of the error log file.
# If you do not specify an ErrorLog directive within a <VirtualHost>
# container, error messages relating to that virtual host will be
# logged here.  If you *do* define an error logfile for a <VirtualHost>
# container, that host's errors will be logged there and not here.
#
ErrorLog ${APACHE_LOG_DIR}/error.log

#
# LogLevel: Control the severity of messages logged to the error_log.
# Available values: trace8, ..., trace1, debug, info, notice, warn,
# error, crit, alert, emerg.
# It is also possible to configure the log level for particular modules, e.g.
# "LogLevel info ssl:warn"
#
LogLevel warn

# Include module configuration:
IncludeOptional mods-enabled/*.load
IncludeOptional mods-enabled/*.conf

# Include list of ports to listen on
Include ports.conf


# Sets the default security model of the Apache2 HTTPD server. It does
# not allow access to the root filesystem outside of /usr/share and /var/www.
# The former is used by web applications packaged in Debian,
# the latter may be used for local directories served by the web server. If
# your system is serving content from a sub-directory in /srv you must allow
# access here, or in any related virtual host.
<Directory />
	Options FollowSymLinks
	AllowOverride None
	Require all denied
</Directory>

<Directory /usr/share>
	AllowOverride None
	Require all granted
</Directory>

<Directory /var/www/>
	Options Indexes FollowSymLinks
	AllowOverride None
	Require all granted
</Directory>

#<Directory /srv/>
#	Options Indexes FollowSymLinks
#	AllowOverride None
#	Require all granted
#</Directory>




# AccessFileName: The name of the file to look for in each directory
# for additional configuration directives.  See also the AllowOverride
# directive.
#
AccessFileName .htaccess

#
# The following lines prevent .htaccess and .htpasswd files from being
# viewed by Web clients.
#
<FilesMatch "^\.ht">
	Require all denied
</FilesMatch>


#
# The following directives define some format nicknames for use with
# a CustomLog directive.
#
# These deviate from the Common Log Format definitions in that they use %O
# (the actual bytes sent including headers) instead of %b (the size of the
# requested file), because the latter makes it impossible to detect partial
# requests.
#
# Note that the use of %{X-Forwarded-For}i instead of %h is not recommended.
# Use mod_remoteip instead.
#
LogFormat "%v:%p %h %l %u %t \"%r\" %>s %O \"%{Referer}i\" \"%{User-Agent}i\"" vhost_combined
LogFormat "%h %l %u %t \"%r\" %>s %O \"%{Referer}i\" \"%{User-Agent}i\"" combined
LogFormat "%h %l %u %t \"%r\" %>s %O" common
LogFormat "%{Referer}i -> %U" referer
LogFormat "%{User-agent}i" agent

# Include of directories ignores editors' and dpkg's backup files,
# see README.Debian for details.

# Include generic snippets of statements
IncludeOptional conf-enabled/*.conf

# Include the virtual host configurations:
IncludeOptional sites-enabled/*.conf

# vim: syntax=apache ts=4 sw=4 sts=4 sr noet

 sites-enabled/000-default.conf :

<VirtualHost *:80>
	# The ServerName directive sets the request scheme, hostname and port that
	# the server uses to identify itself. This is used when creating
	# redirection URLs. In the context of virtual hosts, the ServerName
	# specifies what hostname must appear in the request's Host: header to
	# match this virtual host. For the default virtual host (this file) this
	# value is not decisive as it is used as a last resort host regardless.
	# However, you must set it for any further virtual host explicitly.
	#ServerName www.example.com

	ServerAdmin webmaster@localhost
	DocumentRoot /var/www/html

	# Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
	# error, crit, alert, emerg.
	# It is also possible to configure the loglevel for particular
	# modules, e.g.
	#LogLevel info ssl:warn

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

	# For most configuration files from conf-available/, which are
	# enabled or disabled at a global level, it is possible to
	# include a line for only one particular virtual host. For example the
	# following line enables the CGI configuration for this host only
	# after it has been globally disabled with "a2disconf".
	#Include conf-available/serve-cgi-bin.conf
</VirtualHost>

# vim: syntax=apache ts=4 sw=4 sts=4 sr noet

**Comment by jdieter** [comment]
Just wanted to bring this issue up-to-date.  I've setup my test apache server identically to the configuration sent to me privately by @sbabic, but still can't reproduce.

@sbabic, what OS are you running zchunk on?  And, since you're talking about OE systems, is the OS 32-bit or 64-bit? (Just wondering if we're hitting some kind of integer wrap-around here)

**Comment by jdieter** [comment]
Nevermind, it's not integer wrap-around.  I've just re-read comment 3.  If it's happening in curl, it's definitely the web server.

**Comment by jdieter** [comment]
@sbabic, just wanting to clarify something.  When I test the curl command you used in comment 3, I end up with a 301845 byte file (which is what I'd expect).  How big is the file you are getting?

**Comment by jdieter** [comment]
As verified by testing, fixed in 54665404f3747fcc91e8000d5c34d60ba2342c3f.
